### PR TITLE
[77] Improve test suite for RSAEncrypter/Decrypter

### PR DESCRIPTION
**Pull request by gigi-mohemian**
Resolves part of issue #77 by extending the RSA encrypter/decrypter test suite.

The test keypair is extended into a set of two keypairs, one for `Alice` and one for `Bob`. Additional tests are added to test encryption and decryption of payloads with these keys. Some tests are added for specific error cases, i.e. trying to decrypt with the wrong key.

Looking forward to your comments! 💻 👀

### CODE REVIEW

File: Tests/RSADecrypterTests.swift [hunk 1]
```diff
@@ -25,7 +25,11 @@ import XCTest
...
+    let cipherTextWithAliceKeyBase64URL = "RHLdaxC52ve0icBLGh9Xn_jNoV3uJlEuLRer8_7OOqY2QCGdWPb8OiuxxjEep8okfvlrw_eheEll6v8xdjpzP2zopVEBoNZ9J8BDKL1Y-tTi176jlGXBXKJdRitsp14sVHx7x-dW9FjiOHEn_CSm3bM2THAJOEtjjeF6vRvxpGxm1KruTbtZ37VTl-vM-e4STda5dCanRPKFemDPYQvjUAF_wuE-P-UeE0fpxDceuCRN3C-H8LL9TIrSETRV_CzPVf6Ki9zJpaZQAnqZh5ix0KhaHFMawV7TJetcKPbCzEHbxAF5ib16mxKFc5Th3QFS3eRKYxjaEdciXWVMCrGfXQ"
+    let cipherTextWithBobKeyBase64URL = "TwSS87gjm4cbFbda5klqW_Io2oCoo2zAxWJaU6L1jAeGEoL6u1Gw7vPdEzGGNvUliqsdBBCe8yQhnft9nqIZ9CD7enfVz6tFWvEJuVZbhy0eZad_LLFBFm8eDEgmlw7n7nAZxeww0WloffMmK6GPUE5SjQSWjCRAYA9xwacbxM8shH6hIwpelwV3f-QJr6M_hoAVkiqZJ2tFfiEciPhVVuqNZxxudMxaNj-bWfSipwh2T11kxYQD3EdVR4MDkyxz2JTUQc20OIf_F4WwQXJlbVNLTs4miDKHmw7Ddd45D2DJs_nH-bMQK8b6iH3nke15arP23rozuagF4SHhenApPw"
+    let cipherTextWithAliceKeyEmptyStringBase64URL = "hyOA/tmOclFkj31UPrRb1EnaRMhR5VZg5TrUyfLMtCUlh3grAva0+sSjqt6zSlWK06A6zUieV69aLRbJ0ZactTTqX2CFrhiZ5nUXhzuUya83VKBI0xrGkpQ8u1y2Iqgb+gbWsFJdJ41cSpZXRpc16Hhd3klTp7YydYZQUG//PLM5bn359kqpT8meJdGqTceehVxmdqTpVwukh/uqLOE8CBCrT7D/2t18mzApGpm/Su4bVbZggJ5g9MRPSnwgq1GjKNcMa1PKd+/OWB/rIeDmorT8dLrusGeLbwFCj1HEz4z5izamiBiyPh96G0m4ZhPtVhR4Fo3ARj9C037GroDQ2w=="
```
Comment (carol-mohemian) [inline Tests/RSADecrypterTests.swift]: Is this base64URL? 🤔 I guess the padding(`==`) and the `/` isn't 😁

File: .gitignore [hunk 2]
```diff
@@ -101,3 +101,4 @@ fastlane/test_output
 
 .idea/
 sonar-reports/
+/JOSESwift.xcworkspace/xcshareddata
```
Comment (daniel-mohemian) [inline .gitignore]: 🙏

File: Tests/RSADecrypterTests.swift (hunk 1 above)
Comment (daniel-mohemian) [reply Tests/RSADecrypterTests.swift]: Nope, you're right. It's not base64url. @gigi-mohemian you can quickly convert it online [here](https://cryptii.com) or by hand replace all `/` with `_`, all `+` with `-` and remove the trailing `==`.

Or you just remove the `URL` from the variable name. 😄

Comment (daniel-mohemian) [reply Tests/RSADecrypterTests.swift]: Are the cipher texts are generated by openssl?

I know the original string was a nastly long thing as well, but now that we started touching the tests again, what do you think of staying below 120 characters per line (if possible). With strings its kind of easy to use [Swift's multiline string literals](https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/StringsAndCharacters.html):

e.g.:

``` swift
let cipherText = """
    asdf\
    jklö
"""

File: Tests/RSADecrypterTests.swift [hunk 3]
```diff
@@ -35,28 +39,81 @@ class RSADecrypterTests: CryptoTestCase {
...
 
-        let decrypter = RSADecrypter(algorithm: .RSA1_5, privateKey: privateKey2048!)
-        let plainText = try! decrypter.decrypt(Data(base64URLEncoded: cipherTextBase64URL)!)
+        let decrypter = RSADecrypter(algorithm: .RSA1_5, privateKey: privateKeyBob2048!)
+        let plainText = try! decrypter.decrypt(Data(base64URLEncoded: cipherTextWithBobKeyBase64URL)!)
 
         XCTAssertEqual(plainText, message.data(using: .utf8))
     }
 
+    func testDecryptingAliceSecretWithBobKey() {
```
Comment (daniel-mohemian) [inline Tests/RSADecrypterTests.swift]: 👌

**Comment by gigi-mohemian** [comment]
@carol-mohemian @daniel-mohemian I addressed the issues mentioned. Please have a look again 👍

File: Tests/RSADecrypterTests.swift [hunk 4]
```diff
@@ -25,9 +25,24 @@ import XCTest
...
-    let cipherTextWithAliceKeyEmptyStringBase64URL = "hyOA/tmOclFkj31UPrRb1EnaRMhR5VZg5TrUyfLMtCUlh3grAva0+sSjqt6zSlWK06A6zUieV69aLRbJ0ZactTTqX2CFrhiZ5nUXhzuUya83VKBI0xrGkpQ8u1y2Iqgb+gbWsFJdJ41cSpZXRpc16Hhd3klTp7YydYZQUG//PLM5bn359kqpT8meJdGqTceehVxmdqTpVwukh/uqLOE8CBCrT7D/2t18mzApGpm/Su4bVbZggJ5g9MRPSnwgq1GjKNcMa1PKd+/OWB/rIeDmorT8dLrusGeLbwFCj1HEz4z5izamiBiyPh96G0m4ZhPtVhR4Fo3ARj9C037GroDQ2w=="
+
+    let cipherTextWithAliceKeyBase64 = """
+RHLdaxC52ve0icBLGh9Xn_jNoV3uJlEuLRer8_7OOqY2QCGdWPb8OiuxxjEep8okfvlrw_eheEll6v8xdjpzP2zopVEBoNZ9J8BDKL1Y-tTi176jlGXBXKJ\
+dRitsp14sVHx7x-dW9FjiOHEn_CSm3bM2THAJOEtjjeF6vRvxpGxm1KruTbtZ37VTl-vM-e4STda5dCanRPKFemDPYQvjUAF_wuE-P-UeE0fpxDceuCRN3C\
+-H8LL9TIrSETRV_CzPVf6Ki9zJpaZQAnqZh5ix0KhaHFMawV7TJetcKPbCzEHbxAF5ib16mxKFc5Th3QFS3eRKYxjaEdciXWVMCrGfXQ
+"""
+
+    let cipherTextWithBobKeyBase64 = """
```
Comment (carol-mohemian) [inline Tests/RSADecrypterTests.swift]: Hmm, now this is a little confusing, why is `cipherTextWithBobKeyBase64` and cipherTextWithAliceKeyBase64` base64URL and `cipherTextWithAliceKeyEmptyStringBase64` "only" base64?

Comment (gigi-mohemian) [reply Tests/RSADecrypterTests.swift]: Ah damn, you're right. I'll clean this up asap.

**Comment by gigi-mohemian** [comment]
Sorry this took so long @daniel-mohemian @carol-mohemian ⏳ 

The tests for RSA encryption/decryption should be fine now, I re-created the cipher texts using Alice's and Bob's keys from scratch using `openssl`, since it's been so long 😅I wanted to make sure that everything is correct.

I'd like to do additional tests for AES and for the signer/verifier, but I think it's best to do it in another PR.

Please have another look and let me know what you think 😘

File: Tests/RSADecrypterTests.swift (hunk 4 above)
Comment (gigi-mohemian) [reply Tests/RSADecrypterTests.swift]: It's base64 encoded now.

```
printf "The true sign of intelligence is not knowledge but imagination." | openssl rsautl -encrypt -pubin -inkey alice.pub.pem -out >(base64)
```

**Comment by gigi-mohemian** [comment]
I agree as well! 😄 

Double-checked the RFC and updated the tests. Added the following:

*Encryption:*

- test if encryption of maximum length message `mLen` works
- test if encryption fo `mLen + 1` fails

*Decryption:*

- test if decryption of `len = k` works (i.e., does not throw a "length" error)
- test if decryption of `len = k-1` fails
- test if decryption of `len = k+1` fails

File: Tests/RSADecrypterTests.swift [hunk 5]
```diff
@@ -131,4 +131,56 @@ sVyhqpFuZQ6hhklG9lJr6OBBuk/+pcJYdHuYEuLnJhPeKqF/9xgMOU0e0xLMtkQW+IfDMlm0oAVavHrx
...
+        // https://tools.ietf.org/html/rfc3447#section-7.2.2
+        let cipherTextLengthInBytes = SecKeyGetBlockSize(privateKeyAlice2048!)
+        let testMessage = Data(count: cipherTextLengthInBytes)
+
+        let decrypter = RSADecrypter(algorithm: .RSA1_5, privateKey: privateKeyAlice2048!)
+
+        XCTAssertThrowsError(try decrypter.decrypt(testMessage)) { (error: Error) in
+            // Should throw "decryption failed", but
+            // should _not_ throw cipherTextLenghtNotSatisfied
+            XCTAssertNotEqual(error as? RSAError, RSAError.cipherTextLenghtNotSatisfied)
```
Comment (carol-mohemian) [inline Tests/RSADecrypterTests.swift]: This test will fail as soon as #96 is done 🤔

File: Tests/RSADecrypterTests.swift [hunk 6]
```diff
@@ -131,4 +131,56 @@ sVyhqpFuZQ6hhklG9lJr6OBBuk/+pcJYdHuYEuLnJhPeKqF/9xgMOU0e0xLMtkQW+IfDMlm0oAVavHrx
...
+            XCTFail()
+            return
+        }
+
+        let cipherTextLengthInBytes = SecKeyGetBlockSize(privateKeyAlice2048!)
+        let testMessage = Data(count: cipherTextLengthInBytes + 1)
+
+        let decrypter = RSADecrypter(algorithm: .RSA1_5, privateKey: privateKeyAlice2048!)
+        XCTAssertThrowsError(try decrypter.decrypt(testMessage)) { (error: Error) in
+            XCTAssertEqual(error as? RSAError, RSAError.cipherTextLenghtNotSatisfied)
```
Comment (carol-mohemian) [inline Tests/RSADecrypterTests.swift]: Same #96 (just to give a heads up)

File: Tests/RSADecrypterTests.swift (hunk 5 above)
Comment (daniel-mohemian) [reply Tests/RSADecrypterTests.swift]: Good catch!

**Pull request merged**